### ISSUE TITLE
some questions about the implement's details

### ISSUE THREAD

**Issue by Interesting6** [dragen1860/MAML-Pytorch#23]
First, thanks for your easily understand coding work.

Question1:
in your meta.py program:
`
            for k in range(1, self.update_step):
                # 1. run the i-th task and compute loss for k=1~K-1
                logits = self.net(x_spt[i], fast_weights, bn_training=True)
                loss = F.cross_entropy(logits, y_spt[i])
                # 2. compute grad on theta_pi
                grad = torch.autograd.grad(loss, fast_weights)
                # 3. theta_pi = theta_pi - train_lr * grad
                fast_weights = list(map(lambda p: p[1] - self.update_lr * p[0], zip(grad, fast_weights)))

                logits_q = self.net(x_qry[i], fast_weights, bn_training=True)
                # loss_q will be overwritten and just keep the loss_q on last update step.
                loss_q = F.cross_entropy(logits_q, y_qry[i])
                losses_q[k + 1] += loss_q

                with torch.no_grad():
                    pred_q = F.softmax(logits_q, dim=1).argmax(dim=1)
                    correct = torch.eq(pred_q, y_qry[i]).sum().item()  # convert to numpy
                    corrects[k + 1] = corrects[k + 1] + correct
`
why this fast_weight backpropogation part required to be uodated update_step times?

Question2:
in omniglotNShot.py, function load_data_cache()'s definition,
`for sample in range(10):  # num of episodes`
it's that means an episode contains batchse(here is 32) tasks?
aren't a task equal to an episode?

I see, in function next(), I can't understand why you set:
`if self.indexes[mode] >= len(self.datasets_cache[mode]):
            self.indexes[mode] = 0
            self.datasets_cache[mode] = self.load_data_cache(self.datasets[mode])`
I think it means when epoch turn to the multiple of 10, function next() will generate 10 times of 32 tasks, and every epoch use one(32 tasks) of them(10)?
so why not remove it, and remove the 'for' loop that above mentioned, just use function load_data_cache() in next() function?

Question3:
every time using fine-tuning aren't producing a new fine-tuned model?
I thought it was just using support set to calculate loss then backpropagation, which results in the fine-tuned model then evaluation in query set.
I see it use the same update parameters strategy as meta procedure, and just calculate the accuracy of the query set.

**Comment by eugval** [comment]
Hey!
My understanding is that:
Question1:
With MAML you can choose to do one inner loop gradient update or multiple ones, which is what the ```update_step``` loop here represents. In the reported results, you also get the accuracies you obtain after allowing one, two three ect gradient updates.

Question2: Not sure, haven't looked at the data generation very closely.

Question3 : It is not to ruin the batch_norm state which calculates the mean and variance of the hidden layers on forward passes as well.
